doc/gnoi-reboot/gnoi_reboot.md

```diff
@@ -0,0 +1,252 @@
+
+The Framework container is envisioned to provide a home for foundational services that interact with multiple containers/applications.
+
+State verification daemon periodically (or by request) asks applications to verify their state. For example:
+
+
+
+* SWSS Cfgmgr confirms CONFIG DB matches APPL DB.
+* Syncd verifies ASIC DB matches SDK cache.
+
+State verification daemon aggregates application results into a single state and writes overall state to the STATE DB.
+
+It is proposed to create a new “Framework” container with reboot as its first daemon and state verification as its second.
+
+
+
+![drawing](./images/image1.png)
+
+One step further: the Reboot BE Daemon can have multiple state machines for different subcomponents. For example:
+
+
+
+* System reboot
+* Optic reboot
+* Container reboot
+
+
+
+
+
+
+![drawing](./images/image2.png)
+
+
+## High-Level Design
+
+
+
+
+
+![drawing](./images/image4.png)
+
+The gNOI System service ([system.proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto)) defines Reboot, RebootStatus and CancelReboot rpc’s. These rpc’s can be used to reboot a switch (or subcomponent) via gNOI.
+
+
+### Telemetry &lt;-> Reboot BE API
+
+There are three reboot rpc’s defined in [system.proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto): Reboot, RebootStatus and CancelReboot.
+
+JSON encoded RebootRequest, RebootStatusRequest and CancelRebootRequest will be send to the Reboot BE via a redis notification channel named: “`Reboot_Request_Channel".`
+
+JSON encoded RebootResponse, RebootStatusResponse and CancelRebootResponse will be send to the gNOI Reboot FE  via a redis notification channel named: “`Reboot_Response_Channel".`
+
+
+### Reboot Request Notification Schema
+
+Data is JSON encoded [RebootRequest](https://github.com/openconfig/gnoi/blob/98d6b81c6dfe3c5c400209f82a228cf3393ac923/system/system.proto#L103C13-L103C13), [RebootStatusRequest](https://github.com/openconfig/gnoi/blob/98d6b81c6dfe3c5c400209f82a228cf3393ac923/system/system.proto#L145) or [CancelRebootRequest](https://github.com/openconfig/gnoi/blob/98d6b81c6dfe3c5c400209f82a228cf3393ac923/system/system.proto#L137) (from [system.proto](http://google3/third_party/openconfig/gnoi/system/system.proto)) indicated by the Key value.
+
+
+### Reboot Response Notification Schema
+
+Where StatusCode is defined in [sonic-swss-common/common/status_code_util.h](https://github.com/sonic-net/sonic-swss-common/blob/master/common/status_code_util.h) and can be populated via swss::statusCodeToStr(StatusCode::SWSS_RC_SUCCESS).
+
+Data is JSON encoded RebootRequestResponse, RebootStatusResponse or CancelRebootResponse (from [system.proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto)) indicated by the Key value.
+
+
+### CLI
+
+A cli based tool/script can write/read to/from the “Reboot_Request_Channel” and “Reboot_Response_Channel” notification channels to provide cli based reboot support.
+
```

Comment:
Could you provide a sample of what this CLI would look like (and how it's distinguished from the preexisting `reboot` CLI)? Including the option of specifying subcomponents?